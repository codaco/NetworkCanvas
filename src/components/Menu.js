import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { MenuContent } from '.';
import { MenuItem } from './Elements';

/**
  * Renders a menu, updating styles on DOM elements outside of this.
  * @extends Component
  */
class MenuFactory extends Component {
  /**
    * adds listener for key events to close Menu
    */
  componentDidMount() {
    window.onkeydown = this.listenForClose;
  }

  /**
    * adds listener for key events to close Menu
    */
  componentWillUnmount() {
    window.onkeydown = null;
  }

  listenForClose = (e) => {
    const event = e || window.event;

    if (this.props.isOpen && (event.key === 'Escape' || event.keyCode === 27)) {
      this.props.toggleMenu();
    }
  }

  // intercepts click events; calls callback and toggles Menu open state
  menuItemClick = (itemClick) => {
    itemClick();
    this.props.toggleMenu();
  }

  render() {
    const { isOpen, items, toggleMenu, searchField } = this.props;

    const menuItems = items.map(item =>
      (<MenuItem
        key={item.id}
        to={item.to}
        onClick={() => this.menuItemClick(item.onClick)}
        title={item.title}
        isActive={item.isActive}
        interfaceType={item.interfaceType}
      />),
    );

    return (
      <div className="menu">
        <div className={isOpen ? 'menu__wrap menu__wrap--isOpen' : 'menu__wrap'}>
          <div className="menu__content">
            <MenuContent
              toggleMenu={toggleMenu}
              searchField={searchField}
              items={menuItems}
            />
          </div>
        </div>
        {!isOpen && <div className="menu__burger">
          <button onClick={toggleMenu} className="ui large button">
            <i className="download icon" />
            Burger
          </button>
        </div>}
      </div>
    );
  } // end render
} // end class

MenuFactory.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  items: PropTypes.array,
  toggleMenu: PropTypes.func.isRequired,
  searchField: PropTypes.object,
};

MenuFactory.defaultProps = {
  items: [],
  searchField: null,
};

export default MenuFactory;
