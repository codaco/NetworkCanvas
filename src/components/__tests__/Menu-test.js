/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import Menu from '../Menu';

const items = [{ id: 'a', title: 'a title', to: 'a-to' },
          { id: 'b', title: 'b title', to: 'b-to' },
          { id: 'c', title: 'c title', to: 'c-to' }];

describe('Menu component', () => {
  it('renders menu with list', () => {
    const component = renderer.create(
      <Menu items={items} isOpen={false} toggleMenu={() => {}} />,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  // TODO flush out tests, but the Menu is calling document.getElementById instead of using refs
});
