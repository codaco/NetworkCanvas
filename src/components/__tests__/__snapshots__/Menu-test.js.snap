exports[`Menu component renders menu with list 1`] = `
<div
  className="menu">
  <div
    className="menu__wrap">
    <div
      className="menu__content">
      <div
        className="scrollable">
        <div
          className="scrollable__window">
          <div>
            <div
              className="menu__cross">
              <button
                className="ui large button"
                onClick={[Function]}>
                <i
                  className="download icon" />
                Cross
              </button>
            </div>
            <header>
              <h1
                className="menu__title">
                Stages
              </h1>
            </header>
            <nav>
              <a
                className="menu__menuitem menu__menuitem--"
                onClick={[Function]}
                role="menuitem"
                tabIndex={0}>
                a title
              </a>
              <a
                className="menu__menuitem menu__menuitem--"
                onClick={[Function]}
                role="menuitem"
                tabIndex={0}>
                b title
              </a>
              <a
                className="menu__menuitem menu__menuitem--"
                onClick={[Function]}
                role="menuitem"
                tabIndex={0}>
                c title
              </a>
            </nav>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="menu__burger">
    <button
      className="ui large button"
      onClick={[Function]}>
      <i
        className="download icon" />
      Burger
    </button>
  </div>
</div>
`;
